### PR TITLE
Enable fp8 inference for Llava-hf 7B and 13B in 1.16 release

### PR DESCRIPTION
**Pull request by Luca-Calabria**
# What does this PR do?

Reference link [GS-119](https://habana.atlassian.net/browse/GS-119)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/habana/transformers/models/clip/modeling_clip.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import torch
+from transformers.models.clip.modeling_clip import CLIPVisionEmbeddings
+
+
+class GaudiCLIPVisionEmbeddings(CLIPVisionEmbeddings):
+    def forward(self, pixel_values: torch.FloatTensor) -> torch.Tensor:
+        batch_size = pixel_values.shape[0]
+        target_dtype = self.patch_embedding.weight.dtype
+        if "float8" in str(target_dtype):
```
Comment (schoi-habana) [inline optimum/habana/transformers/models/clip/modeling_clip.py:9]: add a condition to check  self.patch_embedding.weight/pixel_values are on hpu

**Review by mandy-li — CHANGES_REQUESTED** [review]
please also test llava-13b for both bf16/fp8 and add the tests to CI

File: tests/test_image_to_text_example.py [hunk 2]
```diff
@@ -16,6 +16,9 @@
         "bf16": [
             ("llava-hf/llava-1.5-7b-hf", 1, 87.2901500056982),
         ],
+        "fp8": [
+            ("llava-hf/llava-1.5-7b-hf", 1, 123.00953973789325),
```
Comment (schoi-habana) [inline tests/test_image_to_text_example.py:20]: i got 2x performance for both bf16 and fp8 using docker 1.16.0-447. can you confirm?

Comment (Luca-Calabria) [reply tests/test_image_to_text_example.py:20]: this was tested on 1.16.0-388
I'll test also on your version

**Comment by mandy-li** [comment]
@Luca-Calabria , one more: please add the command examples to README. thanks

File: tests/test_image_to_text_example.py (hunk 2 above)
Comment (Luca-Calabria) [reply tests/test_image_to_text_example.py:20]: It seems latest available on Gaudi devcloud is 1.16.0-410

**Comment by Luca-Calabria** [comment]
> please also test llava-13b for both bf16/fp8 and add the tests to CI

@mandy-li  Done

**Comment by Luca-Calabria** [comment]
> @Luca-Calabria , one more: please add the command examples to README. thanks

Done

File: optimum/habana/transformers/models/clip/modeling_clip.py (hunk 1 above)
Comment (Luca-Calabria) [reply optimum/habana/transformers/models/clip/modeling_clip.py:9]: Done

File: tests/test_image_to_text_example.py (hunk 2 above)
Comment (Luca-Calabria) [reply tests/test_image_to_text_example.py:21]: @schoi-habana can you please try to run your test using **--max_new_tokens 20** ? If I remove it, so it uses its default value 100, then the Throughput is 2x as you pointed out.

Comment (schoi-habana) [reply tests/test_image_to_text_example.py:21]: right i was not using --max_new_tokens 20. i got about same performance as you with --max_new_tokens 20

File: examples/image-to-text/README.md [hunk 3]
```diff
@@ -30,4 +30,27 @@ python3 run_pipeline.py \
...
-  - [Salesforce/blip-image-captioning-base](https://huggingface.co/Salesforce/blip-image-captioning-base)
\ No newline at end of file
+  - [Salesforce/blip-image-captioning-base](https://huggingface.co/Salesforce/blip-image-captioning-base)
+
+## Single-HPU FP8 inference
+
+2-steps run:  
+1) Measuring Model
+```bash
+QUANT_CONFIG="<full-parent-path>/optimum-habana/examples/image-to-text/quantization_config/maxabs_measure.json" python3 run_pipeline.py \
```
Comment (mandy-li) [inline examples/image-to-text/README.md:40]: follow the format in text-generation README. do not provide command that can't be run by copy/paste

File: optimum/habana/transformers/models/clip/modeling_clip.py (hunk 1 above)
Comment (schoi-habana) [reply optimum/habana/transformers/models/clip/modeling_clip.py:15]: i meant 
if "float8" in str(target_dtype) and pixel_values.device.type == "hpu":

File: examples/image-to-text/README.md [hunk 4]
```diff
@@ -30,4 +30,27 @@ python3 run_pipeline.py \
...
+    --model_name_or_path llava-hf/llava-1.5-7b-hf \
+    --image_path "https://llava-vl.github.io/static/images/view.jpg" \
+    --use_hpu_graphs \
+    --bf16 \
+    --quant_config
+```
+
+2) Quantization Model Inference
+```bash
+QUANT_CONFIG="<full-parent-path>/optimum-habana/examples/image-to-text/quantization_config/maxabs_quant.json" python3 run_pipeline.py \
```
Comment (mandy-li) [inline examples/image-to-text/README.md:50]: same comment as measurement

File: examples/image-to-text/README.md [hunk 5]
```diff
@@ -30,4 +30,27 @@ python3 run_pipeline.py \
...
+    --model_name_or_path llava-hf/llava-1.5-7b-hf \
+    --image_path "https://llava-vl.github.io/static/images/view.jpg" \
+    --use_hpu_graphs \
+    --bf16 \
+    --quant_config
```
Comment (mandy-li) [inline examples/image-to-text/README.md:55]: remove

File: examples/image-to-text/run_pipeline.py [hunk 6]
```diff
@@ -69,6 +69,7 @@ def main():
         action="store_true",
         help="Whether to perform generation in bf16 precision.",
     )
+    parser.add_argument("--quant_config", action="store_true", help="Enable Quantization to fp8")
```
Comment (mandy-li) [inline examples/image-to-text/run_pipeline.py:72]: remove

**Review by mandy-li — CHANGES_REQUESTED** [review]

File: examples/image-to-text/run_pipeline.py (hunk 6 above)
Comment (Luca-Calabria) [reply examples/image-to-text/run_pipeline.py:72]: It seems this flag https://github.com/HabanaAI/optimum-habana-fork/blob/habana-main/tests/test_text_generation_example.py#L139 confused me.
The expected format is to enable args.quant_config if the QUANT_CONFIG env is set up. Got it

File: optimum/habana/transformers/models/clip/modeling_clip.py (hunk 1 above)
Comment (Luca-Calabria) [reply optimum/habana/transformers/models/clip/modeling_clip.py]: Updated

File: examples/image-to-text/README.md (hunk 3 above)
Comment (Luca-Calabria) [reply examples/image-to-text/README.md]: Done

File: examples/image-to-text/README.md (hunk 4 above)
Comment (Luca-Calabria) [reply examples/image-to-text/README.md]: Done

File: examples/image-to-text/README.md (hunk 5 above)
Comment (Luca-Calabria) [reply examples/image-to-text/README.md]: Done

File: examples/image-to-text/run_pipeline.py (hunk 6 above)
Comment (Luca-Calabria) [reply examples/image-to-text/run_pipeline.py]: Done

**Review by mandy-li — CHANGES_REQUESTED** [review]
minor changes needed

File: examples/image-to-text/README.md [hunk 7]
```diff
@@ -30,4 +30,29 @@ python3 run_pipeline.py \
...
+Here is an example to measure the tensor quantization statistics on Llava-1.5-7b:
+```bash
+QUANT_CONFIG=./quantization_config/maxabs_measure.json python run_pipeline.py \
+--model_name_or_path llava-hf/llava-1.5-7b-hf \
+--image_path "https://llava-vl.github.io/static/images/view.jpg" \
+--use_hpu_graphs \
+--bf16
+```
+
+Here is an example to quantize the model based on previous measurements for LLama2-70b:
```
[the remark this answers is not in the document]
Comment (Luca-Calabria) [reply examples/image-to-text/README.md]: my fault, sorry.
Fixed

**Comment by mandy-li** [comment]
@Luca-Calabria , LGTM. Did you run `make style` to check code format?

**Comment by Luca-Calabria** [comment]
> @Luca-Calabria , LGTM. Did you run `make style` to check code format?

I do it now

**Comment by Luca-Calabria** [comment]
@mandy-li I did "make style" commit

**Review by mandy-li — APPROVED** [review]

**Comment by mandy-li** [comment]
LGTM. @regisss , can you review pls? also i don't see any reasons why the unit tests fail.

**Comment by regisss** [comment]
> LGTM. @regisss , can you review pls? also i don't see any reasons why the unit tests fail.

Unit tests fail because I have to actually trigger them for people who are not collaborators of this repo.

@Luca-Calabria I tried the measure command in the README with SynapseAI v1.15.1 and got this error:
```
AttributeError: 'PatchedKVCache' object has no attribute 'quant_input_0'
```
It's supposed to work with v1.16 I guess?

**Comment by Luca-Calabria** [comment]
> > LGTM. @regisss , can you review pls? also i don't see any reasons why the unit tests fail.
> 
> Unit tests fail because I have to actually trigger them for people who are not collaborators of this repo.
> 
> @Luca-Calabria I tried the measure command in the README with SynapseAI v1.15.1 and got this error:
> 
> ```
> AttributeError: 'PatchedKVCache' object has no attribute 'quant_input_0'
> ```
> 
> It's supposed to work with v1.16 I guess?

@regisss exactly, you must test it on version 1.16

**Review by regisss — APPROVED** [review]

**Pull request merged**